tiledb/sm/query/query_condition.cc

```diff
@@ -0,0 +1,468 @@
+      return *static_cast<const T*>(cell_value) !=
+             *static_cast<const T*>(clause.condition_value_);
+    default:
+      LOG_FATAL("Unknown query condition op");
+  }
+
+  return false;
+}
+
+bool QueryCondition::cmp_clause_str(
+    void* const cell_value,
+    const uint64_t cell_value_size,
+    const Clause& clause) const {
+  std::string cell_value_str(
+      static_cast<const char*>(cell_value), cell_value_size);
+  std::string condition_value_str(
+      static_cast<const char*>(clause.condition_value_),
+      clause.condition_value_size_);
+
+  switch (clause.op_) {
+    case QueryConditionOp::LT:
+      return cell_value_str < condition_value_str;
+    case QueryConditionOp::LE:
+      return cell_value_str <= condition_value_str;
+    case QueryConditionOp::GT:
+      return cell_value_str > condition_value_str;
+    case QueryConditionOp::GE:
+      return cell_value_str >= condition_value_str;
+    case QueryConditionOp::EQ:
+      return cell_value_str == condition_value_str;
+    case QueryConditionOp::NE:
+      return cell_value_str != condition_value_str;
+    default:
+      LOG_FATAL("Unknown query condition op");
+  }
+
+  return false;
+}
+
+template <class T>
+Status QueryCondition::cmp_clause(
+    ResultTile* const result_tile,
+    const uint64_t cell_idx,
+    const Clause& clause,
+    bool* const cmp) const {
+  T cell_value;
+  RETURN_NOT_OK(
+      result_tile->read(clause.field_name_, &cell_value, 0, cell_idx, 1));
+  *cmp = cmp_clause<T>(&cell_value, clause);
+  return Status::Ok();
+}
+
+Status QueryCondition::cmp_clause_str(
+    ResultTile* const result_tile,
+    const uint64_t cell_idx,
+    const Clause& clause,
+    bool* const cmp) const {
+  const Tile& tile = std::get<0>(*result_tile->tile_tuple(clause.field_name_));
+  const uint64_t cell_size = tile.cell_size();
+
+  char* const cell_value = static_cast<char*>(tdb_malloc(cell_size));
+
+  const Status st =
+      result_tile->read(clause.field_name_, cell_value, 0, cell_idx, 1);
+  if (!st.ok()) {
+    tdb_free(cell_value);
+    return st;
+  }
+
+  *cmp = cmp_clause_str(cell_value, cell_size, clause);
+
+  tdb_free(cell_value);
+
+  return Status::Ok();
+}
+
+Status QueryCondition::cmp_cell(
+    const ArraySchema* const array_schema,
+    ResultTile* const result_tile,
+    const uint64_t cell_idx,
+    bool* const cmp) const {
+  if (clauses_.empty() || result_tile->cell_num() == 0) {
+    *cmp = true;
+    return Status::Ok();
+  }
+
+  // We have a combination op between each clause, where each
+  // combination op is an 'AND'.
+  assert(clauses_.size() - 1 == combination_ops_.size());
+
+  for (const auto& clause : clauses_) {
+    const Attribute* const attribute =
+        array_schema->attribute(clause.field_name_);
+    if (!attribute) {
+      return LOG_STATUS(Status::QueryConditionError(
+          "Unknown attribute " + clause.field_name_));
+    }
+
+    const Datatype type = attribute->type();
+    switch (type) {
```

Comment:
Similar to what is done here for the datatype, we should be doing also for the operands and use functors that implement tight loops over the attribute buffers for vectorized computation. The result should be a bytemap. The bytemap should then be used to join results across clauses, and skip clauses if the bytemap is all 0.